include/deal.II/numerics/nonlinear.h

```diff
@@ -0,0 +1,401 @@
+   * initial guess, and stores the final solution in the same vector.
+   */
+  void
+  solve(VectorType &initial_guess_and_solution);
+
+  std::function<void(VectorType &)> reinit_vector;
+
+  std::function<int(const VectorType &src, VectorType &dst)> residual;
+
+  std::function<int(const VectorType &src, VectorType &dst)>
+    iteration_function;
+
+  std::function<int(const VectorType &current_u, const VectorType &current_f)>
+    setup_jacobian;
+
+  std::function<
+    int(const VectorType &rhs, VectorType &dst, const double tolerance)>
+    solve_with_jacobian;
+
+protected:
+
+  AdditionalData additional_data;
+
+  std::string type;
+
+  // default is 1e-8, no particular reason
+  double tolerance = 1e-8;
+
+private:
+
+#ifdef DEAL_II_WITH_SUNDIALS
+typename SUNDIALS::KINSOL<VectorType>::AdditionalData
+        additional_data_kinsol;
+#endif
+
+#ifdef DEAL_II_WITH_TRILINOS
+  typename TrilinosWrappers::NOXSolver<VectorType>::AdditionalData additional_data_nox;
+  Teuchos::RCP<Teuchos::ParameterList> parameters_nox = Teuchos::rcp(new Teuchos::ParameterList);
+#endif
+      
+};
+
+template<typename VectorType>
+NLSolverSelector<VectorType>::NLSolverSelector()
+{}
+
+template<typename VectorType>
+NLSolverSelector<VectorType>::NLSolverSelector(const AdditionalData &additional_data)
+  : additional_data(additional_data)
+{
+  #ifdef DEAL_II_WITH_SUNDIALS
+  // These if statements pass on the strategy to the other nonlinear solvers
+  // There might be a better way to do this...
+  if (additional_data.strategy == NLSolverSelector<VectorType>::AdditionalData::linesearch)
+    additional_data_kinsol.strategy = SUNDIALS::KINSOL<VectorType>::AdditionalData::linesearch;
+  else if (additional_data.strategy == NLSolverSelector<VectorType>::AdditionalData::newton)
+    additional_data_kinsol.strategy = SUNDIALS::KINSOL<VectorType>::AdditionalData::newton;
+  else if (additional_data.strategy == NLSolverSelector<VectorType>::AdditionalData::picard)
+    additional_data_kinsol.strategy = SUNDIALS::KINSOL<VectorType>::AdditionalData::picard;
+
+  additional_data_kinsol.maximum_non_linear_iterations = additional_data.maximum_non_linear_iterations;
+  additional_data_kinsol.function_tolerance = additional_data.function_tolerance;
+  additional_data_kinsol.step_tolerance = additional_data.step_tolerance;
+  additional_data_kinsol.no_init_setup = additional_data.no_init_setup;
+  additional_data_kinsol.maximum_setup_calls = additional_data.maximum_setup_calls;
+  additional_data_kinsol.maximum_newton_step = additional_data.maximum_newton_step;
+  additional_data_kinsol.dq_relative_error = additional_data.dq_relative_error;
+  additional_data_kinsol.maximum_beta_failures = additional_data.maximum_beta_failures;
+  additional_data_kinsol.anderson_subspace_size = additional_data.anderson_subspace_size;
+  #endif
+
+  #ifdef DEAL_II_WITH_TRILINOS
+  additional_data_nox.max_iter = additional_data.maximum_non_linear_iterations;
```

Comment:
How do you choose between Newton and Picard?

Reply:
You choose between the two the same way you do KINSOL, with NOX you can't really choose between Newton and Picard so that option is chosen elsewhere in the NOX parameter specifications.